src/context/WalletProvider/KeepKeyProvider.tsx

```diff
@@ -0,0 +1,157 @@
+    label: ['walletProvider.keepKey.settings.descriptions.timeoutDuration', { minutes: '15' }]
+  },
+  {
+    value: DeviceTimeout.TwentyMinutes,
+    label: ['walletProvider.keepKey.settings.descriptions.timeoutDuration', { minutes: '20' }]
+  },
+  {
+    value: DeviceTimeout.ThirtyMinutes,
+    label: ['walletProvider.keepKey.settings.descriptions.timeoutDuration', { minutes: '30' }]
+  },
+  {
+    value: DeviceTimeout.FortyFiveMinutes,
+    label: ['walletProvider.keepKey.settings.descriptions.timeoutDuration', { minutes: '45' }]
+  },
+  {
+    value: DeviceTimeout.SixtyMinutes,
+    label: ['walletProvider.keepKey.settings.descriptions.timeoutDuration', { minutes: '60' }]
+  }
+]
+
+export enum KeepKeyActions {
+  SET_HAS_PIN_CACHING = 'SET_HAS_PIN_CACHING',
+  SET_HAS_PASSPHRASE = 'SET_HAS_PASSPHRASE',
+  SET_DEVICE_TIMEOUT = 'SET_DEVICE_TIMEOUT',
+  RESET_STATE = 'RESET_STATE'
+}
+
+export interface InitialState {
+  hasPinCaching: boolean | undefined
+  hasPassphrase: boolean | undefined
+  keepKeyWallet: KeepKeyHDWallet | undefined
+  deviceTimeout: RadioOption<DeviceTimeout> | undefined
+}
+
+const initialState: InitialState = {
+  hasPinCaching: undefined,
+  hasPassphrase: undefined,
+  keepKeyWallet: undefined,
+  deviceTimeout: timeoutOptions[0]
+}
+
+export interface IKeepKeyContext {
+  state: InitialState
+  setHasPinCaching: (enabled: boolean) => void
+  setHasPassphrase: (enabled: boolean) => void
+  keepKeyWallet: KeepKeyHDWallet | undefined
+}
+
+export type KeepKeyActionTypes =
+  | { type: KeepKeyActions.SET_HAS_PIN_CACHING; payload: boolean | undefined }
+  | { type: KeepKeyActions.SET_HAS_PASSPHRASE; payload: boolean | undefined }
+  | { type: KeepKeyActions.SET_DEVICE_TIMEOUT; payload: RadioOption<DeviceTimeout> | undefined }
+  | { type: KeepKeyActions.RESET_STATE }
+
+const reducer = (state: InitialState, action: KeepKeyActionTypes) => {
+  switch (action.type) {
+    case KeepKeyActions.SET_HAS_PASSPHRASE:
+      return { ...state, hasPassphrase: action.payload }
+    case KeepKeyActions.SET_HAS_PIN_CACHING:
+      return { ...state, hasPinCaching: action.payload }
+    case KeepKeyActions.SET_DEVICE_TIMEOUT:
+      return { ...state, deviceTimeout: action.payload }
+    case KeepKeyActions.RESET_STATE:
+      return initialState
+    default:
+      return state
+  }
+}
+
+const KeepKeyContext = createContext<IKeepKeyContext | null>(null)
+
+export const KeepKeyProvider = ({ children }: { children: React.ReactNode }): JSX.Element => {
+  const {
+    state: { wallet }
+  } = useWallet()
+  const keepKeyWallet = useMemo(() => (wallet && isKeepKey(wallet) ? wallet : undefined), [wallet])
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  const setHasPinCaching = useCallback((payload: boolean | undefined) => {
+    dispatch({
+      type: KeepKeyActions.SET_HAS_PIN_CACHING,
+      payload
+    })
+  }, [])
+
+  const setHasPassphrase = useCallback((payload: boolean | undefined) => {
```

Comment:
what benefit are you getting from `useCallback`?

Reply:
The `useEffect` below depends on `setHasPassphrase` (as well as `setDeviceTimeout` and `setHasPinCaching`).

If these functions aren't wrapped by `useCallback` they'll be recreated every render, creating an infinite render loop (or at least exceeding the render stack depth limit).

TL;DR, without it the app crashes and we get a white screen.